---
title: "dagLogo Vignette"
author: "Jianhong Ou, Lihua Julie Zhu"
date: "`r doc_date()`"
package: "`r pkg_ver('dagLogo')`"
bibliography: bibliography.bib
abstract: >
  A sequence logo has been widely used as a graphical representation of an alignment of multiple amino acid (AA) or nucleic acid sequences. Our R/bioconductor package dagLogo can be used to visualize amio acids of significantly differential usage, with or without grounping based on their physical and chemical properties, among given sequence contexts compared to a background context.
vignette: >
  %\VignetteIndexEntry{dagLogo Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document
---

```{r echo=FALSE, results='hide', warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
    library(dagLogo)
    library(biomaRt)
    library(UniProt.ws)
    library(motifStack)
    library(Biostrings)
    library(grDevices)
    })
```

#Introduction
A sequence logo has been widely used as a graphical representation of an alignment of multiple amino acid (AA) or nucleic acid sequences.
There is a package seqlogo[@Oliver2006] implemented in R to draw DNA sequence logos. 
And another package motifStack[@Jianhong2012] was developed for drawing sequence logos for amino Acid, DNA and RNA sequences. motifStack also has the capability for graphical representation of multiple motifs.

IceLogo[@Colaert2009] is a tool developed in Java to visualize significantly conserved sequence patterns in
an alignment of multiple peptide sequences against background sequences. Compare to 
webLogo[@Crooks2004], which relying on information theory, iceLogo builds on probability theory. It is
reported that iceLogo has a more dynamic nature and is more appropriate for analysis of
conserved sequence patterns.

However, iceLogo can only compare conserved sequences to reference sequences at the individual amino acid level. 
As we know, some conserved sequence patterns are not conserved at the individual amino acid level, but conserved at the level of amino acids grouped on the basis of their physical and chemical properties, such as charge, hydrophobicity and etc.

Here we developed a R/Bioconductor package dagLogo, inspired by iceLogo, to visualize significantly conserved sequence patterns relative to a proper background set of sequences, with or without grouping amino acid residuals based on their physical and chemical properties. The flowchart of performaing dagLogo analysis is shown as [Figure 1]{./dagLogo_flowchart.png}

#Set up the working environment
To visualize the result of statistical test of the differential usage of grouped amino acid, the full path to the executable of GhostScript need to be set by using the environment variable R\_GSCMD per session or by adding it to the system environment variable PATH. For a Unix, Linux or Mac operating system (OS), "gs" is used for searching the PATH variable, while For a Windows OS the environment variable "GSC" is searched for first, and then "gswi64c.exe" or "gswin32c.exe" are looked for among the system environment variables, depending on the operating system.

For example on a Windowsplatform with gswin32c.exe installed at C:\\ Program Files\\ gs\\ gs9.06\\ bin, you can set up R\_GSCMD as follows:
```{r setenv, eval=FALSE, echo=TRUE}
Sys.setenv(R_GSCMD = file.path("C:", "Program Files", "gs", 
                               "gs9.06", "bin", "gswin32c.exe"))
```

#Demonstration of dagLogo application

## Step 1: Fetching peptide sequences from the Biomart
### Case 1: You have both positions of the anchoring AAs and the identifiers of their enclosing peptide/protein sequences for fetching sequences using the fetchSequence function via the Biomart.
```{r fetchSequences1}
##just in case biomaRt server does not response
try({
    mart <- useMart("ensembl")
    fly_mart <-
        useDataset(mart = mart, dataset = "dmelanogaster_gene_ensembl")
    dat <- read.csv(system.file("extdata", "dagLogoTestData.csv",
                                package = "dagLogo"))
    dat <- dat[1:5, ] ##subset to speed sample
    seq <- fetchSequence(
        IDs = as.character(dat$entrez_geneid),
        anchorPos = as.character(dat$NCBI_site),
        mart = fly_mart,
        upstreamOffset = 7,
        downstreamOffset = 7
    )
    head(seq@peptides)
})
```

### Case 2: You don't have the exactly postion infor, but You have the interesting peptide subsequences and the identifiers of their enclosing peptide/protein sequences for fetching sequences using the fetchSequence function via the Biomart. In the following examples, the anchoring AAs are marked by asterisks.

```{r fetchSequences2}
try({
    mart <- useMart("ensembl")
    fly_mart <-
        useDataset(mart = mart, dataset = "dmelanogaster_gene_ensembl")
    dat <- read.csv(system.file("extdata", "dagLogoTestData.csv",
                                package = "dagLogo"))
    seq <- fetchSequence(
        IDs = as.character(dat$entrez_geneid),
        anchorAA = "*",
        anchorPos = as.character(dat$peptide),
        mart = fly_mart,
        upstreamOffset = 7,
        downstreamOffset = 7
    )
    head(seq@peptides)
})
```

In following example, the anchoring AAs are lower case "s" for amino acid serine.

```{r fetchSequences3}
if(interactive()){
    try({
            dat <- read.csv(system.file("extdata", "peptides4dagLogo.csv",
                                package = "dagLogo"))
            mart <- useMart("ensembl")
            human_mart <-
            useDataset(mart = mart, dataset = "hsapiens_gene_ensembl")
            seq <- fetchSequence(IDs = toupper(as.character(dat$symbol)),
                                type = "hgnc_symbol",
                                anchorAA = "s",
                                anchorPos = as.character(dat$peptides),
                                mart = human_mart,
                                upstreamOffset = 7,
                                downstreamOffset = 7)
            head(seq@peptides)
    })
}
```
Similarly, peptide sequences can be fected from an predefined Proteome object.  
  
  
### Case 3: You already have the aligned peptides sequences in hand. Then you can use the formatSequence function to prepare an object of dagPeptides. Befor doing that, you need prepare a Proteome object by the prepareProteome function.

```{r formatSequence}
dat <- unlist(read.delim(system.file("extdata", "grB.txt", package = "dagLogo"),
                        header = F, as.is = TRUE))
head(dat)
##prepare proteome from a fasta file
proteome <- prepareProteome(fasta = system.file("extdata",
                                                "HUMAN.fasta",
                                                package = "dagLogo"), 
                            species = "Homo sapiens")
##prepare an object of dagPeptides
seq <- formatSequence(seq = dat, proteome = proteome, upstreamOffset = 14,
                      downstreamOffset = 15)
```

## Step 2: Building background models  
Once you have an object of dagPeptides in hand, you can start to build a background model
for DAG test. The background could be random subsequences of a whole proteome or your inputs.
If the background was built from a whole proteome or proteome without your inputs, 
an object of Proteome is required. 

Sequences provided by a fasta file or downloaded from the UniProt database can be used to prepare a Proteome object.
Case 3 of Step 1 shows how to prepare a Proteome object from a fasta file. Here we show how to prepare an object of
Proteome via the UniProt database.

```{r prepareProteome0}
if(interactive()) {
    proteome <- prepareProteome(source = "UniProt", species = "Homo sapiens")
}
```

Then the proteome is used for background model building for Fisher's exact test or *Z*-test as follows. 

```{r buildBackgroundModel}
bg_fisher <- buildBackgroundModel(seq, background = "wholeProteome", proteome = proteome, testType = "fisher")
bg_ztest <- buildBackgroundModel(seq, background = "wholeProteome", proteome = proteome, testType = "ztest")
```

## Step 3: Testing differential usage of (grouped) amino acids
Test can be done without making any change to the formatted, aligned amino acid symbols. Alternatively, amino acids can be grouped on the basis of their physical and chemical properties and the formatted, aligned amino acid symbols are replaced by new sets of symbols for each group, such as "P", "N" and "U" for amino acids with positive charge, negative charge and no charge if the amino acids are grouped based on their charge.  

```{r testDAU}
## no grouping

t0 <- testDAU(seq, dagBackground = bg_ztest)
## grouded classically: nonpolar_aliphatic = c("A", "G", "L", "M", "I", "V"), polar_uncharged = c("C", "P", "Q", "S", "T"), aromatic = c("F", "W", "Y"), positively_charged = c("H", "K", "N", "R"), negatively_charged = c("D", "E").
t1 <- testDAU(dagPeptides = seq, dagBackground = bg_ztest, groupingScheme = "classic")

## grouded on the basis of charge: positive = c("H", "K", "R"), neutral = c("A", "C", "F", "G", "I", "L", "M", "N", "P", "Q", "S", "T", "V","W","Y"), negative = c("D", "E")
t2 <- testDAU(dagPeptides = seq, dagBackground = bg_ztest, groupingScheme = "charge")

## grouped on the basis of their chemical property: hydrophobic = c("A", "F", "I", "L", "M", "P", "V", "W"), polar = c("C", "G", "S", "T", "Y"), basic = c("H", "K", "R"), neutral = c("N", "Q"), acidic = c("D", "E")
t3 <- testDAU(dagPeptides = seq, dagBackground = bg_ztest, groupingScheme = "chemistry")

## grouped on the basis of hydrophobicity: hydrophilic = c("D", "E", "K", "N", "Q", "R"), neutral = c("A", "G", "H", "P", "S", "T"), hydrophobic = c("C", "F", "I", "L", "M", "V", "W", "Y")
t4 <- testDAU(dagPeptides = seq, dagBackground = bg_ztest, groupingScheme = "hydrophobicity")
```

## Step 4: Visualizing test results
We can use a heatmap or logo to display the test results.

```{r dagHeatmap,fig.cap="DAG heatmap",fig.width=8,fig.height=6}
##Plot a heatmap to show the results
dagHeatmap(t0) 
```

```{r dagLogo0,fig.cap="ungrouped results",fig.width=8,fig.height=6}
##Plot a logo to show the ungrouped results
dagLogo(t0) 
```

```{r dagLogo1,fig.cap="classic grouped",fig.width=8,fig.height=6}
##Plot a logo to show the classic grouped results
dagLogo(t1, groupingSymbol = getGroupingSymbol(t1@group), legend = TRUE)
```

```{r dagLogo2,fig.cap="charge grouped",fig.width=8,fig.height=6}
##Plot a logo to show the charge grouped results
dagLogo(t2, groupingSymbol = getGroupingSymbol(t2@group), legend = TRUE)
```

```{r dagLogo3,fig.cap="chemistry grouped",fig.width=8,fig.height=6}
##Plot a logo to show the chemistry grouped results
dagLogo(t3, groupingSymbol = getGroupingSymbol(t3@group), legend = TRUE)
```

```{r dagLogo4,fig.cap="hydrophobicity grouped",fig.width=8,fig.height=6}
##Plot a logo to show the hydrophobicity grouped results
dagLogo(t4, groupingSymbol = getGroupingSymbol(t4@group), legend = TRUE)
```

#Another full example: Comparative analysis of the DNA-binding motif of CAPs using the packages motifStack and dagLogo  

Catabolite Activator Protein (CAP), also known as cAMP Receptor Protein (CRP),is a transcription activator that binds more than 100 sites of the *E. coli* genome.

The motif of the DNA-binding helix-turn-helix motif of the CAP family is first visualized by using the motifStack package.  Residues 7-13 form the first helix, 14-17 the turn and 18-26 the DNA recognition helix. The glycine at position 15 appears to be critical in forming the turn.

```{r CAPmotif,fig.cap="Catobolite Activator Protein Motif",fig.width=8,fig.height=6}
protein <- read.table(file.path(find.package("motifStack"), "extdata", "cap.txt"))
protein <- t(protein[, 1:20])
motif <- pcm2pfm(protein)
motif <- new("pfm", mat = motif,name = "CAP",
            color = colorset(alphabet = "AA", colorScheme = "chemistry"))
##The DNA-binding helix-turn-helix motif of the CAP family ploted by motifStack
plotMotifLogo(motif@mat, motifName = motif@name,
              p = motif@background, colset = motif@color)
```

For comparison, then the CAP DNA-binding motif is visualized using the dagLogo package.  
 
```{r CAPdagLogo,fig.cap="Catobolite Activator Protein Motif",fig.width=8,fig.height=6}
cap <- as.character(readAAStringSet(system.file("extdata", 
                                                "cap.fasta", 
                                                package="dagLogo")))
seq <- formatSequence(seq = cap, proteome = ecoli.proteome)
bg <- buildBackgroundModel(seq,
                           background= "wholeProteome",
                           proteome = ecoli.proteome,
                           numSubsamples = 10L)
##The DNA-binding helix-turn-helix motif of the CAP family ploted by dagLogo
t0 <- testDAU(seq, bg)
dagLogo(t0)
```

Residuals at positions 10, 14, 16, 21 and 25 are partially or completely buried in the 3-D structure of CAPs are preferentially hydrophobic. Thee amino acids of the CAP DNA-binding motifs are grouped on the basis of their chemical proterties and visualized using dagLogo as follows. This plot clearly displays those preferentially hydrophobic sites.
```{r CAPgroup,fig.cap="Catobolite Activator Protein Motif",fig.width=8,fig.height=6}
## The DNA-binding helix-turn-helix motif of the CAP family grouped by chemistry
t1 <- testDAU(seq, bg, groupingScheme = "chemistry")
dagLogo(t1, groupingSymbol = getGroupingSymbol(t1@group), legend = TRUE)
```

# Session Info
```{r sessionInfo, echo=TRUE}
sessionInfo()
```

